### PR TITLE
Add alterType and update index.d.ts for alter function

### PR DESCRIPTION
**Pull request by OlivierCavadenti**
- Fix https://github.com/knex/knex/pull/4730/files#r791157891
- And I think we are good for an 1.0.2 (doc here : https://github.com/knex/documentation/pull/385)

### CODE REVIEW

**Comment by OlivierCavadenti** [comment]
@intech do you think I can do something for cockroachdb test ? Or I need to revert for now ?

**Comment by intech** [comment]
> @intech do you think I can do something for cockroachdb test ? Or I need to revert for now ?

Quickly studied the error and can not say how to get around it. I'll ask the cockroachdb engineers for advice and get back.

**Comment by intech** [comment]
@OlivierCavadenti This column change is not yet supported https://github.com/cockroachdb/cockroach/issues/47636
And https://github.com/cockroachdb/cockroach/issues/49329#issuecomment-1022120446 for other test case

**Comment by kibertoad** [comment]
wait, wasn't the whole idea of the change to avoid altering nullability? or crdb doesn't support such change even without toucing nullability?

**Comment by intech** [comment]
@kibertoad I don't really understand the real application of this functionality, which I wrote about in the original issue on another project using knex.
Now the error occurs on this request:
```sql
alter table "primary_table_null" alter column "id_not_nullable_primary" type integer using ("id_not_nullable_primary"::integer);
```
Why do we need to change the column type to int if we already have int when creating the table?
```sql
create table "primary_table_null" ("id_not_nullable_primary" integer not null);
```

@OlivierCavadenti @kibertoad Could you please explain to me the application of these queries?

**Comment by kibertoad** [comment]
@intech I think what happens there is that since knex does not know what is the current type of the column, it uses the one specified in alter command. To address that we probably need additional parameter alterType, which omits type change if set.

File: types/index.d.ts [hunk 1]
```diff
@@ -2210,7 +2210,7 @@ export declare namespace Knex {
     notNullable(): ColumnBuilder;
     nullable(): ColumnBuilder;
     comment(value: string): ColumnBuilder;
-    alter(): ColumnBuilder;
+    alter(options: Readonly<{alterNullable?: boolean}>): ColumnBuilder;
```
Comment (kibertoad) [inline types/index.d.ts:2213]: looks like we need another option `alterType`

**Review by kibertoad — APPROVED** [review]

File: types/index.d.ts [hunk 2]
```diff
@@ -2210,7 +2210,7 @@ export declare namespace Knex {
     notNullable(): ColumnBuilder;
     nullable(): ColumnBuilder;
     comment(value: string): ColumnBuilder;
-    alter(): ColumnBuilder;
+    alter(options: Readonly<{alterNullable?: boolean, alterType?: boolean}>): ColumnBuilder;
```
Comment (B4nan) [inline types/index.d.ts:2213]: this is breaking for TS users. the parameter was not there, now its required.

Comment (OlivierCavadenti) [reply types/index.d.ts:2213]: Yeah, sorry, just fixed and merged here : https://github.com/knex/knex/pull/4996

Comment (kibertoad) [reply types/index.d.ts:2213]: will publish hotfix shortly

**Pull request merged**